### ISSUE TITLE
Datastore initialization fails because one model do not sync

### ISSUE THREAD

**Issue by meducati** [aws-amplify/amplify-js#6691]
**Describe the bug**
This is driving me crazy.... 

Following the event order from Datastore start as described here: https://docs.amplify.aws/lib/datastore/datastore-events/q/platform/js

For some reason, one of my models (1 of 8) do not sync. This issue only apples on one my three ENVs (same app, different backends). The data in this model (compared to its dev + test counterparts) is slightly bigger, but only 462 rows.

Datastore emits no errors for me to catch.- Setting debug level to WARN shows "Utils - error on" (pic below). Consequently, I never get the "ready" event either.

I assume this relates to DynamoDB throttling, but my database is small (abt 1000 rows in total), however with three M:M relationship tables.

**Expected behavior**
All models should sync, or at least emit an error event (with a good explaining why the error occurred).

**Screenshots**
Setting warn levels to window.LOG_LEVEL = 'DEBUG' provides:

![utils error](https://user-images.githubusercontent.com/63409162/91640500-51c4ec80-ea1e-11ea-9762-c82b6fff4c3b.JPG)
(Successful result from model 7 + "silent" Utils error)
As seen in the picture, the data from the 462 items have been downloaded from the cloud correctly, so the issue must be with initializing the model locally. (Datastore has correctly added the table in indexeddb, but it is not populating it with the data).

![sdkclientexception](https://user-images.githubusercontent.com/63409162/91640568-db74ba00-ea1e-11ea-935a-9b321148c347.JPG)
(A second error appears at the same location. This happen about 3 seconds after the page has reloaded, so the error message regarding 10000 milliseconds cannot be accurate).

![hot partition](https://user-images.githubusercontent.com/63409162/91696147-767ca980-eb6f-11ea-8a55-06e6479216f8.JPG)
(A few times, I also got this error...)

<details>
  <summary><strong>Environment</strong></summary>

```
System:
    OS: Windows 7 6.1.7601
    CPU: (8) x64 Intel(R) Core(TM) i7-2600 CPU @ 3.40GHz
    Memory: 5.93 GB / 15.98 GB
  Binaries:
    Node: 12.18.3 - C:\Program Files\nodejs\node.EXE
    npm: 6.14.8 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Internet Explorer: 11.0.9600.18921
  npmPackages:
    @aws-amplify/api: ^3.1.22 => 3.1.22
    @aws-amplify/auth: ^3.3.4 => 3.3.4
    @aws-amplify/datastore: ^2.2.9 => 2.2.9
    @aws-amplify/pubsub: ^3.0.23 => 3.0.23
    aws-sdk: ^2.736.0 => 2.736.0
    clean-webpack-plugin: ^3.0.0 => 3.0.0
    copy-webpack-plugin: ^5.1.1 => 5.1.1
    css-loader: ^3.5.3 => 3.6.0
    html-webpack-plugin: ^3.2.0 => 3.2.0
    ini: ^1.3.5 => 1.3.5
    inquirer: ^6.5.1 => 6.5.2
    save-dev: 0.0.1-security => 0.0.1-security
    style-loader: ^1.2.1 => 1.2.1
    terser-webpack-plugin: ^4.1.0 => 4.1.0
    webpack: ^4.44.1 => 4.44.1
    webpack-cli: ^3.3.12 => 3.3.12
    webpack-dev-server: ^3.11.0 => 3.11.0
    webpack-merge: ^5.1.1 => 5.1.2
    webpack-pwa-manifest: ^4.1.1 => 4.2.0
    workbox-webpack-plugin: ^5.1.3 => 5.1.3
  npmGlobalPackages:
    @aws-amplify/api: 3.1.10
    @aws-amplify/auth: 3.2.7
    @aws-amplify/cli: 4.27.0
    @aws-amplify/core: 2.2.4
    @aws-amplify/datastore: 1.0.7
    @aws-amplify/pubsub: 3.0.11
    http-server: 0.12.1
    npm: 6.14.8
    webpack-cli: 3.3.11
    webpack-dev-server: 3.10.3
    webpack: 4.41.6
```

</details>

**Additional context**
Schema:

type BatchFile
@model(queries:null)
@auth(rules: [
    {allow: owner},
    {allow: groups, groups: ["admin"] },
    {allow: groups, groups: ["users"] },
    {allow: private, provider: iam }
  ])
@key(name: "filesByBatch", fields: ["batchId", "fileId"])
@key(name: "batchesByFile", fields: ["fileId", "batchId"])
{
  id: ID!
  batchId: ID!
  fileId: ID!
  batch: CADBatch @connection(fields: ["batchId"])
  file: CADFile @connection(fields: ["fileId"])
}

**Comment by meducati** [comment]
Follow up:
To circumvent this issue until I find a solution, whenever I normally would use "Database.query", I now have to await and check if the array returns empty. If it is empty, I do an API.GraphQL call, using the Amplify-constructed "syncQuery", which return all my 462 entries correctly.

It feels like this is not related to the DB size, but rather a conflict somewhere. Because we are on the production env, I am a bit reluctant to mess with the API. What is safe to try without risk of losing any cloud data?

**Comment by amhinson** [comment]
@meducati I see you are using older versions of the `@aws-amplify` packages. Would you be able to try with `aws-amplify@latest` to see if you're still experiencing the same issues?

You also only need `aws-amplify` as a dependency in your project instead of each Amplify package independently. Having mismatched versions can potentially cause other unexpected issues.

**Comment by meducati** [comment]
@amhinson 

Thank you for replying (much appreciated). I have followed you advice and my new environment is as follows:

<details>
  <summary><strong>Environment</strong></summary>



```
System:
    OS: Windows 7 6.1.7601
    CPU: (8) x64 Intel(R) Core(TM) i7-2600 CPU @ 3.40GHz
    Memory: 4.75 GB / 15.98 GB
  Binaries:
    Node: 12.18.3 - C:\Program Files\nodejs\node.EXE
    npm: 6.14.8 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Internet Explorer: 11.0.9600.18921
  npmPackages:
    @aws-amplify/datastore: ^2.5.1 => 2.5.1
    aws-amplify: ^3.3.1 => 3.3.1
    aws-sdk: ^2.736.0 => 2.736.0
    clean-webpack-plugin: ^3.0.0 => 3.0.0
    copy-webpack-plugin: ^6.1.1 => 6.1.1
    css-loader: ^3.5.3 => 3.6.0
    html-webpack-plugin: ^3.2.0 => 3.2.0
    ini: ^1.3.5 => 1.3.5
    inquirer: ^6.5.1 => 6.5.2
    save-dev: 0.0.1-security => 0.0.1-security
    style-loader: ^1.2.1 => 1.2.1
    terser-webpack-plugin: ^4.1.0 => 4.1.0
    webpack: ^4.44.1 => 4.44.1
    webpack-cli: ^3.3.12 => 3.3.12
    webpack-dev-server: ^3.11.0 => 3.11.0
    webpack-merge: ^5.1.1 => 5.1.2
    webpack-pwa-manifest: ^4.1.1 => 4.2.0
    workbox-webpack-plugin: ^5.1.3 => 5.1.3
  npmGlobalPackages:
    @aws-amplify/cli: 4.29.2
    http-server: 0.12.1
    npm: 6.14.8
    webpack-cli: 3.3.11
    webpack-dev-server: 3.10.3
    webpack: 4.41.6
```

</details>

**Unfortunately the issue persists!**
(and as mentioned earlier... this only applies to one of my 8 Models)

Some notes:
- After uninstalling all amplify packages and reinstalling aws-amplify, the @aws-datastore folder (+dependencies) were never added into the node_modules folder. Therefore I had to reinstall that one as a separate package.
-  I did clear the entire node_packages folder and "npm install"

![2020-09-24 - Errors1](https://user-images.githubusercontent.com/63409162/94128236-2d600280-fe5a-11ea-8be1-66083ca57d1d.JPG)

Next steps:
- I have yet to populate my test environment with more data than my production environment currently holds. If the issue eventually starts happening in my test environment, I assume the issue is on "my" side. If not, I assume the issue must be with the Cloudformation?

Questions:
- It is still very confusing what these error messages mean. I cannot be hitting a 10 second timeout with DynamoDB


Any other suggestions what I can do to get to the bottom with this would be most welcome!


Sincerely
Bjorn

**Comment by meducati** [comment]
Follow up 2:
I have now populated my TEST environment with more entries and the error is now visible here also. The same model is not being populated. In total, there is now about 2000 entries over 8 models. 4 models are carrying about 450 entries of the data each (including the Model that is not syncing).

**NOTE! As seen in the errors above. ALL entries are being downloaded to the client (syncBatchFiles -> 462 items), but not copied into the Datastore table locally.**

One additional error messages:
- "DynamoDB:SdkClientException"
- "Unable to execute HTTP request: Channel was closed before it could be written to."
- "Unable to execute HTTP request: null"
- "Throughput exceeds the current capacity of your table or index. DynamoDB is automatically scaling your table or index so please try again shortly. If exceptions persist, check if you have a hot key: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/bp-partition-key-d ...

Would be great to get an explanation what actually happens once the data has been downloaded (since it is not being added to Localstore). Are there any capacity limits?

**Comment by amhinson** [comment]
Would you be able to reach out to me (`amhinson`) on our [Discord server](https://discord.gg/jWVbPfC)? It'd be helpful to see more of your app code to debug a bit more.

**Comment by iartemiev** [comment]
@meducati, You should not need to install @aws-amplify/datastore separately. Try removing that from your package.json (aws-amplify should be the only amplify dependency). Then delete your node_modules and package-lock.json
```
rm -rf node_modules package-lock.json
npm install
```

The limit on how much data you can store in IndexedDB (the local storage for DataStore) is quite large. It's very unlikely that you're 
IndexedDB utilizes [Storage API](https://developer.mozilla.org/en-US/docs/Web/API/Storage_API), which in turn can use up to 20% of the total available free storage space. So if your device has 100GB free, 20GB will be shared across all web apps utilizing Storage API. Either way, unless your device is low on free space, there should be multiple GBs available to your IndexedDB store. If you run `await navigator.storage.estimate()` in the Chrome Dev Tools Console, the `quota` parameter will tell you how much space is remaining (in bits).

With regard to the errors you're getting, I have not seen those exact ones before, but my guess is that they might have something to do with either: 
A. The provisioned capacity of your DynamoDB table. Could you try increasing the read/write capacity lets say 2-3x on that table?
B. Malformed/corrupted data being stored in DynamoDB 

Try adjusting your DDB capacity and then after waiting a few minutes for the changes to get applied, run:
```js
await DataStore.clear();
await DataStore.start();
```

One more thing, I don't think this is causing the errors, but I noticed in your schema that you have the following auth rules:
```graphql
@auth(rules: [
  {allow: owner},
  {allow: groups, groups: ["admin"] },
  {allow: groups, groups: ["users"] },
  {allow: private, provider: iam }
])
```
Currently, DataStore does not support combining multiple authorization types on a model. I would suggest removing the last rule `{allow: private, provider: iam }`. You can keep the owner and static group rules, since those all use the same authorization type, Cognito User Pools.

**Comment by iartemiev** [comment]
One more thing - did you happen to manually edit the request/response resolvers for any of your models?

**Comment by meducati** [comment]
> Would you be able to reach out to me (`amhinson`) on our [Discord server](https://discord.gg/jWVbPfC)? It'd be helpful to see more of your app code to debug a bit more.

Thanks. I sent you a message there :)

**Comment by meducati** [comment]
> One more thing - did you happen to manually edit the request/response resolvers for any of your models?

Not the resolvers. I did update the models in schema.js by removing auth ->

`{
                                "groupClaim": "cognito:groups",
                                "provider": "userPools",
                                "allow": "groups",
                                "groups": [
                                    "users"
                                ],
                                "operations": [
                                    "create",
                                    "update",
                                    "delete",
                                    "read"
                                ]
                            }`

This is how I can keep multiple authorization in the backend (for my lambda functions). So, in my amplify project, I have set it up to use both userGroups and IAM. It seems to work (and as mentioned previously... it does work until (presumably) the table gets to large.

**Comment by meducati** [comment]
> @meducati, You should not need to install **@aws-amplify/datastore** separately. Try removing that from your package.json (aws-amplify should be the only amplify dependency). Then delete your node_modules and package-lock.json
> 
> ```
> rm -rf node_modules package-lock.json
> npm install
> ```
> 
> The limit on how much data you can store in IndexedDB (the local storage for DataStore) is quite large. It's very unlikely that you're brushing up against it.
> IndexedDB utilizes [Storage API](https://developer.mozilla.org/en-US/docs/Web/API/Storage_API), which in turn can use up to 20% of the total available free storage space. So if your device has 100GB free, 20GB will be shared across all web apps utilizing Storage API. Either way, unless your device is low on free space, there should be multiple GBs available to your IndexedDB store. If you run `await navigator.storage.estimate()` in the Chrome Dev Tools Console, the `quota` parameter will tell you how much total space you have available (in bits), and `usage` will show how much of that is utilized by the application running in the same tab.
> 
> With regard to the errors you're getting, I have not seen those exact ones before, but my guess is that they might have something to do with either:
> A. The provisioned capacity of your DynamoDB table. Could you try increasing the read/write capacity lets say 2-3x on that table?
> B. Malformed/corrupted data being stored in DynamoDB
> 
> Try adjusting your DDB capacity and then after waiting a few minutes for the changes to get applied, run:
> 
> ```js
> await DataStore.clear();
> await DataStore.start();
> ```
> 
> One more thing, I don't think this is causing the errors, but I noticed in your schema that you have the following auth rules:
> 
> ```graphql
> @auth(rules: [
>   {allow: owner},
>   {allow: groups, groups: ["admin"] },
>   {allow: groups, groups: ["users"] },
>   {allow: private, provider: iam }
> ])
> ```
> 
> Currently, DataStore does not support combining multiple authorization types on a model. I would suggest removing the last rule `{allow: private, provider: iam }`. You can keep the owner and static group rules, since those all use the same authorization type, Cognito User Pools.

Thanks for looking into my issue :)

Some comments:

- I did manage to get datastore automatically installed without a special reference in package.json. One detail less to worry about.
- It is not an storage issue. I already keep track on that :)
- It is unlikely that there is a corruption issue. This model is an M:M model that only contains 3 IDs and 2 connections (as specified in the Amplify Docs. `# M:M relation
# https://aws-amplify.github.io/docs/cli-toolchain/graphql
type BatchFile
@model(queries:null)
@auth(rules: [
    {allow: owner},
    {allow: groups, groups: ["admin"] },
    {allow: groups, groups: ["users"] },
    {allow: private, provider: iam }
  ])
@key(name: "filesByBatch", fields: ["batchId", "fileId"])
@key(name: "batchesByFile", fields: ["fileId", "batchId"])
{
  id: ID!
  batchId: ID!
  fileId: ID!
  batch: CADBatch @connection(fields: ["batchId"])
  file: CADFile @connection(fields: ["fileId"])
}`
... also... the GraphQL alternative works as expected.. and I do see everything that SHOULD go into the local datastore in my error log...
- I do await datastore.clear(), in my code upon user login... I currently do not do datastore.start() but instead rely on it to start when I init Datastore.observe(). I have tried with datastore.start() in the past also.

Untested:
- I am NOT sure on how to increase  read/write capacity since all tables are "on-demand". This could of course be an issue. IF this is an issue, I think it must be because of my poor database design and not the volume of entries:

![schema](https://user-images.githubusercontent.com/63409162/95677548-85229b80-0bc6-11eb-86ae-972fe165e338.png)

Could this schema cause some sort of a roundabout? Increasing datatraffic exponentially? (Again, I do see all records in my console.log, so it must be related to what happens after the data was downloaded, ie. in the sync back to the back-end?)

Side finding:
- If I manually (in schema.js) set syncable to "false", the model does not sync (as supposed), I receive no error (as supposed), but Datastore never fires the "ready" event (so I cannot listen for it and "handle" my issue).

**Comment by meducati** [comment]
Update: Setting ALL models to syncable:false (except BatchFiles), still causes the error.

**Comment by iartemiev** [comment]
@meducati, those errors are being generated by AppSync, not DataStore. They are defined in the Java SDK, for example [`SdkClientException`](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/SdkClientException.html). 

Could you try going to the AppSync Console (you can execute `amplify console api` in terminal) and then running the `syncBatchFiles` query a few times in a row? Can you see the errors then?

Try switching the BatchFile DynamoDB table from On-Demand to provisioned and then let's see if the `ProvisionedThroughputExceededException` starts to show up in your errors. If so, I think you're right about it being an issue with the way schema is modeled.

**Comment by meducati** [comment]
@iartemiev I just played around with Provisioned... First with the default 5,5,5 RCU, WCU and then with 10,10,10. The errors still persist. (No mentioning of ProvisionedThroughputExceededException)

**Comment by iartemiev** [comment]
@meducati did you turn off Auto Scaling as well?

**Comment by meducati** [comment]
Well, I didn´t, but did now... (No change)

In the console (as you suggest), I notice that I can make the tree indefinitely deep since a batch have files, a file has batches, a batch has files and so on....

Is there a limit how deep this goes? (Else it could of course be the reason for timing out).  (I recall there is a default setting within Amplify somewhere, but maybe that is not applicable in the sync?)

Actually, I think this may be the culprit of it all... Thoughts?

**Comment by iartemiev** [comment]
Yes there's a schema max depth that defaults to `2`. So maybe that's why you're able to execute the query via `API.graphql`. 

[Here's](https://docs.amplify.aws/cli/graphql-transformer/codegen#statement-depth) how you can change it after the fact.

But as you suggest, those graphql statements are only utilized when making calls via `API.graphql`. They're not used by DataStore.

**Comment by iartemiev** [comment]
Do you know what is the maximum level of nesting that happens in your data currently? I want to try to reproduce this in my own environment to better understand what's happening. 

Also, if you're able to, could you please post the entire raw output from `Util error on` with all the data and errors? Feel free to upload it here or DM it to me (`@Ivan` in Amplify Staff) on Discord

**Comment by meducati** [comment]
Ivan, my best friend in the world right now! I think we have solved it!!! :)

Updated models:
`type CADBatch @model
@auth(rules: [
    {allow: owner},
    {allow: groups, groups: ["admin"] },
    {allow: groups, groups: ["users"], operations: [read] }
  ])
{
  id: ID!
  status: BatchStatus!    
  name: String!
  createdAt: AWSDateTime
  batchType: BatchType!
  schemaType: SchemaType
  schemaSettings: SchemaSettings
  schemaObjects: [SchemaObject]
  users:[UserBatch] @connection(keyName: "byBatch", fields: ["id"])
  #files: [BatchFile] @connection(keyName: "filesByBatch", fields: ["id"])
}

type BatchFile
@model(queries:null)
@auth(rules: [
    {allow: owner},
    {allow: groups, groups: ["admin"] },
    {allow: groups, groups: ["users"] },
    {allow: private, provider: iam }
  ])
@key(name: "filesByBatch", fields: ["batchId", "fileId"])
@key(name: "batchesByFile", fields: ["fileId", "batchId"])
{
  id: ID!
  batchId: ID!
  fileId: ID!
  #batch: CADBatch @connection(fields: ["batchId"])
  file: CADFile @connection(fields: ["fileId"])
}
`

As you see, I broke the relationship between CADBatch and BatchFile.
- Amplify api push
- Amplify codegen models
(I no longer got an error from having both auth -> userGroups and IAM in the models which is great. No need to edit schema.js anymore)

App now populates all 8 models locally and Datastore fires "ready" upon start.

I do recall that I, a long time ago, increased the "depth" somewhere. Was this in Amplify configure? (This is the only time I ever touched this setting). If you say this is not related to datastore, then it is so. Where can I check the nesting depth?

I will do some additional tests and get back to you on what I would have liked. (Documentation, error messages etc) that would have steered me away from this issue in the first place. (In theory, I still think the way I put up my schema would have been valid since there is not really any "relations" left once you get the data into indexeddb...

I´ll message you on Discord for the error log!

**Comment by iartemiev** [comment]
Great! Please send it over (: I'm definitely interested to see what kind of data you had that was resulting in this issue.

Also, definitely let us know your feedback so that we can improve the docs and error messaging.

**Comment by amhinson** [comment]
@meducati Would you say that the original issue is resolved here and we can close this?

**Comment by meducati** [comment]
> @meducati Would you say that the original issue is resolved here and we can close this?

@amhinson I am still invested in this, but maybe it is time to close this particular issue since I no longer receive the original errors.

**For any late arrivals to this thread, I have been in contact with the Amplify team and in short, here are my concerns.**

- First. I DO have a throttling issue on DynamoDB. The reason for this is probably because how I designed my schema (circular, see earlier post), and how Appsync (not Amplify) handles this.
-- The original errors in my post kind of makes sense now, even though the emitted error descriptions are sub-optimal at best.
-- One would believe that, if my generated graphQL models work as expected, so would Amplify sync (but it is not). My guess is that the reason the GraphQL works, is that it has a depth limit (so my circular database relationships will never go a full "round" and eventually create an infinite loop), but for some reason, this is not the case when Amplify calls Appsync for the sync event.
-- I have tested to "break" one of my relations. At this time the throttling stops.

**Warnings:**
- After updating Amplify CLI to 4.29.2, I no longer receive the error messages in my first post. All models sync and I finally get the "ready" event. However (if I keep all my relations intact), DynamoDB still throttles! What happen is that one of my M:M tables return partial results:
-- All records in my M:M table sync, but not always their nested relations down the tree. Sometimes one of my foreign keys are missing (null) and sometimes it is the other one (inconsistent). (When the foreign key is null, the related table is also null of course). There is NO way to know if this error occurs (as far as I can tell), until your code tries to access this post in a Datastore query).
-- Looking at the DynamoDB console page I also notice that it is not the request to my M:M table that throttle, but the throttle happens on the nested tables. (So this might be the cause for not getting an error client-side)
-- Since the M:M table receive all records (although not the complete set), **Amplify do not raise an error**.
-- The throttling only occurs after you have started adding a bunch of data into DynamoDB, so it can be hard to spot at first.

**Next steps**
I will keep running tests on my schema. Ideally, I would like to keep all my relations.
If I cannot narrow this issue down, my two options are:
- Break relation, re-arrange schema, re-write code (keep Datastore)
- Skip Datastore, and rely on GraphQL calls only (i.e. no offline capabilities)

BIG THANKS to the Amplify team and especially @iartemiev who has provided much insight regarding my issue. Feel free to close this issue now. I will try to post any relevant findings after I have made some more testing.
